### PR TITLE
Meson recognize .C files as C sources while it is C++

### PR DESCRIPTION
**Pull request by jeandet**
This PR adds .C as C++ extension and makes Meson case sensitive for
file extension detection.

I saw that while porting [SAMRAI](https://github.com/jeandet/SAMRAI) to meson they use .C as file extension for all their C++ sources.

### CODE REVIEW

**Comment by nirbheek** [comment]
One thing to keep in mind while doing this is that both macOS (with both HFS+ and APFS, for dumb reasons) and Windows (with both FAT and NTFS because of DOS) use the filesystem in a case-preserving and case-insensitive mode. Even for UTF-8 filenames, which means it's actually locale-dependent.

This means that for the `en_US` locale the following file types are all the same:

```
foo.c
foo.C
FOO.C
FOO.c
```

Note that if you save a file called `foo.C`, it will be listed as `foo.C`, but looking for `foo.C` will also match `foo.c` if it exists.

However! Meson doesn't have to care about any of this in this particular case.

As long as we don't do directory listing, we can require the user to ensure that they're not doing something stupid like having a repository with both `foo.c`/`FOO.c` (C source) and `foo.C`/`FOO.C` (C++ source) in the same directory. Such a thing is impossible on Windows anyway.

I checked SAMRAI and it doesn't do that.

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -637,9 +637,13 @@ def __repr__(self):
...
         if hasattr(src, 'fname'):
             src = src.fname
-        suffix = os.path.splitext(src)[1].lower()
+        # We try first in case sensitive mode detect .C as C++ code instead of C
+        suffix = os.path.splitext(src)[1]
         if suffix and suffix[1:] in self.can_compile_suffixes:
             return True
+        # We keep the old case insensitive behaviour
+        if suffix.lower() and suffix[1:] in self.can_compile_suffixes:
+            return True
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: Why is this needed? We have always required the suffix to be lower case, so there's no old case-insensitive behaviour to keep.

Comment (nirbheek) [reply mesonbuild/compilers/compilers.py]: Proposal: On Linux and BSDs, GCC treats `.C` as a C++ extension, so we should keep the `suffix.lower()` fallback only for other extensions. Looking at the GCC man page, we should also treat `.H` as C++ header for the same reason. The only other extensions that change behaviour in a case-sensitive manner are `.f` and `.F` for fixed-form Fortran and `.s` (asm code) and `.S` (asm code that must be pre-processed).

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -637,9 +637,13 @@ def __repr__(self):
     def can_compile(self, src):
         if hasattr(src, 'fname'):
             src = src.fname
-        suffix = os.path.splitext(src)[1].lower()
+        # We try first in case sensitive mode detect .C as C++ code instead of C
+        suffix = os.path.splitext(src)[1]
         if suffix and suffix[1:] in self.can_compile_suffixes:
             return True
+        # We keep the old case insensitive behaviour
+        if suffix.lower() and suffix[1:] in self.can_compile_suffixes:
```
Comment (textshell) [inline mesonbuild/compilers/compilers.py]: I think this should be `suffix.lower()[1:] in ...`. The two lower at the start of the line doesn't really do anything.

Also the duplicate check for suffix not being empty could be refactored.

Comment (jeandet) [reply mesonbuild/compilers/compilers.py]: Yes here I did a mistake, I don't really understand why ```suffix and suffix[1:]```

File: mesonbuild/compilers/compilers.py (hunk 1 above)
Comment (jeandet) [reply mesonbuild/compilers/compilers.py]: I'm not sure I understand exactly what you want ma to do. 
At least I guess that we agree that we need to do the right language detection from Meson to be able to pass right args?
I know that stupid file systems(case insensitive ones) can make this difficult but it also does for git and many other things.

Comment (nirbheek) [reply mesonbuild/compilers/compilers.py]: What I mean, is that we should not fallback to detecting `.C` as C sources if a C++ compiler is not found. i.e., we should *not* keep the "old case insensitive behaviour" in that case. We should only maintain case-insensitivity for extensions that *are* case-insensitive.

**Comment by nirbheek** [comment]
Closing due to inactivity, let's discuss this in an issue and commit to whether we want to do it at all, before we discuss how to implement it.
